snapshots/windows/windows.go

```diff
@@ -391,3 +410,154 @@ func (s *snapshotter) parentIDsToParentPaths(parentIDs []string) []string {
+	}
+
+	if expand {
+		gbToByte := 1024 * 1024 * 1024
+		if err := hcsshim.ExpandSandboxSize(s.info, filepath.Base(snDir), uint64(gbToByte*sizeGB)); err != nil {
+			return errors.Wrapf(err, "failed to expand sandbox vhdx size to %d GB", sizeGB)
+		}
+	}
+
+	return nil
+}
+
+// This handles creating the UVMs scratch layer. This is a bit different from the containers scratch layer in
+// that expanding the volume doesn't work, so a new set of disks will have to be made. Processing the UVM layer also
+// renders any previously made disks unusable. We keep a record of the last requested scratch size and if it differs,
+// we'll make a new set of vhdx's.
+func (s *snapshotter) createUVMScratchLayer(ctx context.Context, snDir string, parentLayers []string, sizeGB int) error {
+	s.scratchLock.Lock()
+	defer s.scratchLock.Unlock()
+
+	// Hardcoded size of the UVMs scratch in the HCS graphdriver calls. If no size is specified
+	// just keep using the default.
+	if sizeGB == 0 {
+		sizeGB = 10
+	}
+
+	parentLen := len(parentLayers)
+	if parentLen == 0 {
+		return errors.New("no parent layers present")
+	}
+	baseLayer := parentLayers[parentLen-1]
+
+	// Make sure base layer has a UtilityVM folder.
+	uvmPath := filepath.Join(baseLayer, "UtilityVM")
+	if _, err := os.Stat(uvmPath); os.IsNotExist(err) {
+		return errors.Wrapf(err, "failed to find UtilityVM directory in base layer `%s`", baseLayer)
+	}
+
+	var (
+		newSize          bool
+		templateBase     = filepath.Join(uvmPath, "SystemTemplateBase.vhdx")
+		templateDiffDisk = filepath.Join(uvmPath, "SystemTemplate.vhdx")
+		scratchSizeFile  = filepath.Join(uvmPath, "scratchsize")
+	)
+
+	if _, err := os.Stat(scratchSizeFile); os.IsNotExist(err) {
+		// File doesn't exist, this is likely the first time this image is being used. Create the file so
+		// on subsequent runs we can check the last size used. If the size being asked for isn't the default of
+		// 10 then we need to create a new set of disks also.
+		if sizeGB != 10 {
+			newSize = true
+		}
+		if err := ioutil.WriteFile(scratchSizeFile, []byte(fmt.Sprintf("%d", sizeGB)), 0644); err != nil {
+			return errors.Wrap(err, "failed to create a new UVM scratch size file")
+		}
+	} else {
+		data, err := ioutil.ReadFile(scratchSizeFile)
+		if err != nil {
+			return errors.Wrap(err, "failed to read from UVM scratch size file")
+		}
+		if string(data) != fmt.Sprintf("%d", sizeGB) {
+			newSize = true
+			if err := ioutil.WriteFile(scratchSizeFile, []byte(fmt.Sprintf("%d", sizeGB)), 0644); err != nil {
+				return errors.Wrap(err, "failed to write to UVM scratch size file")
+			}
+		}
+	}
+
+	// The ExpandSandboxSize call doesn't work for the UVMs diff disk so we need to create a
+	// new set of disks if we're asking for a new size. Setting up the disks/layer also renders
+	// any previous disks unusable.
+	if newSize {
+		if err := computestorage.SetupUtilityVMBaseLayer(ctx, uvmPath, templateBase, templateDiffDisk, uint64(sizeGB)); err != nil {
```

Comment:
Does this work if we already have another uvm running with this layer? This call will most likely fail in that case.

Also, it seems a bit problematic that we will change the uvm scratch layer size for all the future UVMs based on the request for this current UVM. I wonder if there is a better way to handle this.

Reply:
Removed the uvm scratch size customization for this PR until I can get an answer from the storage folks on why resizing doesn't (seem to) work. Unfortunately as you mentioned offline there's really no way around the other disks becoming unusable after making a new set.